XiEditor/HoverViewController.swift

```diff
@@ -0,0 +1,131 @@
+        self.textColor = NSColor.textColor
+        self.drawsBackground = false
+        self.needsLayout = true
+        self.isVerticallyResizable = true
+        self.alignment = .justified
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+}
+
+class HoverViewController: NSViewController {
+    lazy var scrollView: NSScrollView = {
+        let scrollView = NSScrollView(frame: NSRect(x: 0, y: 0, width: hoverPopoverWidth, height: 0))
+        scrollView.borderType = .noBorder
+        scrollView.hasVerticalScroller = true
+        scrollView.autoresizingMask = [.height]
+        scrollView.contentView.wantsLayer = true
+        scrollView.contentView.layer?.masksToBounds = true
+        scrollView.drawsBackground = false
+        // At most, scroll view will be as tall as its width.
+        scrollView.heightAnchor.constraint(lessThanOrEqualToConstant: hoverPopoverWidth).isActive = true
+        return scrollView
+    }()
+
+    var resultContent: String
+    var hoverView: HoverView
+    let hoverPopoverWidth: CGFloat = 500 // XCode size for quick help popovers
+
+    init(content: String) {
+        self.resultContent = content
+        self.hoverView = HoverView(content: self.resultContent)
+        super.init(nibName: nil, bundle: nil)
+        self.scrollView.documentView = hoverView
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // Required to instantiate view controller programmatically.
+    override func loadView() {
+        self.view = scrollView
+        hoverView.frame.size = scrollView.contentSize
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
+
+    func updateHoverViewColors(newBackgroundColor: NSColor, newTextColor: NSColor) {
+        self.hoverView.backgroundColor = newBackgroundColor
+        self.hoverView.textColor = newTextColor
+    }
+
+    // Returns the height required to fit the hover result.
+    // The text container inset for top and bottom is added to the height required to
+    // draw the string.
+    func heightForContent() -> CGFloat {
+        // Calculates height required for a text view with width = 500.
+        self.hoverView.setFrameSize(NSSize(width: hoverPopoverWidth, height: 0))
+
+        guard let layoutManager = self.hoverView.layoutManager else { return 0 }
+        guard let textContainer = self.hoverView.textContainer else { return 0 }
+
+        layoutManager.glyphRange(for: textContainer)
+        return layoutManager.usedRect(for: textContainer).height + self.hoverView.textContainerInset.height * 2
+    }
+}
+
+extension EditViewController {
+
+    // Puts the popover at the baseline of the chosen hover symbol.
+    func showHover(withResult result: [String: AnyObject]) {
+        if infoPopover.isShown {
+            infoPopover.performClose(self)
+        }
+
+        let hoverContent = result["content"] as! String
```

Comment:
As per latest work in https://github.com/google/xi-editor/pull/753 , the result is now just a String, so you don't need to extract `content` out here.